### DISCUSSION TITLE
Unnecessary INSERT INTO in a query request

### DISCUSSION THREAD

**Discussion by funk1d** [sqlalchemy/sqlalchemy#5855] [Usage Questions] [answered]
**Describe the bug**

Weirdly making an unnecessary `INSERT INTO` in a query request when running the code below.
The second line from last always get executed successfully.
The last line never gets executed successfully, and will be all fine when commented out.
**Expected behavior* 

Just query  without adding `INSERT INTO`.
**To Reproduce**
Please try to provide a [Minimal, Complete, and Verifiable](http://stackoverflow.com/help/mcve) example.
See also [Reporting Bugs](https://www.sqlalchemy.org/participate.html#bugs) on the website, and some [example issues](https://github.com/sqlalchemy/sqlalchemy/issues?q=label%3A%22great+mcve%22).

**Please do not use Flask-SQLAlchemy or any other third-party extensions or dependencies in test cases**.  The test case must illustrate the problem **without** using any third party SQLAlchemy extensions.  Otherwise, please report the bug to those projects first.

```py
from pathlib import Path

from sqlalchemy import Column, ForeignKey, Integer, String, create_engine
from sqlalchemy.engine.base import Engine
from sqlalchemy.ext.declarative import declarative_base
from sqlalchemy.orm import Session, relation

Base = declarative_base()
_engine: Engine = create_engine(
    'sqlite:///' + str(Path(__file__).parent.joinpath('db.sqlite3')),
    echo=True)
db_session = Session(_engine)


class User(Base):
    __tablename__ = 'users'
    _id = Column('id', Integer, primary_key=True)
    _username = Column('username', String, unique=True, nullable=False)
    _user_type_id = Column('user_type_id',
                           Integer,
                           ForeignKey('user_type.id'),
                           nullable=False)
    _user_type = relation('UserType',
                          order_by='UserType._name',
                          backref='_users')

    def __init__(self, username: str, user_type: str):
        self._username = username
        tmp_query = db_session.query(UserType).filter_by(
            _name=user_type).first()  # sqlalchemy makes unnecessary `INSERT INTO` here when executing the last line
        if tmp_query:
            self._user_type = tmp_query
        else:
            self._user_type = UserType(user_type)


class UserType(Base):
    __tablename__ = 'user_type'
    _id = Column('id', Integer, primary_key=True)
    _name = Column('name', String, unique=True, nullable=False)

    def __init__(self, name: str):
        self._name = str(name)


Base.metadata.create_all(_engine)

if not db_session.query(User).filter_by(_username='1').first():
    db_session.add(User('1', '1'))
    db_session.commit()

x = User('1', '1')  # this line can always be executed successfully
x = User('1', '1')  # would be all fine if comment this line
```

**Error**

```
INFO sqlalchemy.engine.base.Engine SELECT CAST('test plain returns' AS VARCHAR(60)) AS anon_1
INFO sqlalchemy.engine.base.Engine ()
INFO sqlalchemy.engine.base.Engine SELECT CAST('test unicode returns' AS VARCHAR(60)) AS anon_1
INFO sqlalchemy.engine.base.Engine ()
INFO sqlalchemy.engine.base.Engine PRAGMA main.table_info("users")
INFO sqlalchemy.engine.base.Engine ()
INFO sqlalchemy.engine.base.Engine PRAGMA main.table_info("user_type")
INFO sqlalchemy.engine.base.Engine ()
INFO sqlalchemy.engine.base.Engine BEGIN (implicit)
INFO sqlalchemy.engine.base.Engine SELECT users.id AS users_id, users.username AS users_username, users.user_type_id AS users_user_type_id 
FROM users 
WHERE users.username = ?
 LIMIT ? OFFSET ?
INFO sqlalchemy.engine.base.Engine ('1', 1, 0)
INFO sqlalchemy.engine.base.Engine SELECT user_type.id AS user_type_id, user_type.name AS user_type_name 
FROM user_type 
WHERE user_type.name = ?
 LIMIT ? OFFSET ?
INFO sqlalchemy.engine.base.Engine ('1', 1, 0)
INFO sqlalchemy.engine.base.Engine INSERT INTO users (username, user_type_id) VALUES (?, ?)
INFO sqlalchemy.engine.base.Engine ('1', 1)
INFO sqlalchemy.engine.base.Engine ROLLBACK
Traceback (most recent call last):
  File "/.venv/lib/python3.9/site-packages/sqlalchemy/engine/base.py", line 1276, in _execute_context
    self.dialect.do_execute(
  File "/.venv/lib/python3.9/site-packages/sqlalchemy/engine/default.py", line 609, in do_execute
    cursor.execute(statement, parameters)
sqlite3.IntegrityError: UNIQUE constraint failed: users.username

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/backend/models.py", line 53, in <module>
    x = User('1', '1')
  File "<string>", line 4, in __init__
  File "/.venv/lib/python3.9/site-packages/sqlalchemy/orm/state.py", line 433, in _initialize_instance
    manager.dispatch.init_failure(self, args, kwargs)
  File "/.venv/lib/python3.9/site-packages/sqlalchemy/util/langhelpers.py", line 68, in __exit__
    compat.raise_(
  File "/.venv/lib/python3.9/site-packages/sqlalchemy/util/compat.py", line 182, in raise_
    raise exception
  File "/.venv/lib/python3.9/site-packages/sqlalchemy/orm/state.py", line 430, in _initialize_instance
    return manager.original_init(*mixed[1:], **kwargs)
  File "/backend/models.py", line 29, in __init__
    tmp_query = db_session.query(UserType).filter_by(
  File "/.venv/lib/python3.9/site-packages/sqlalchemy/orm/query.py", line 3429, in first
    ret = list(self[0:1])
  File "/.venv/lib/python3.9/site-packages/sqlalchemy/orm/query.py", line 3203, in __getitem__
    return list(res)
  File "/.venv/lib/python3.9/site-packages/sqlalchemy/orm/query.py", line 3534, in __iter__
    self.session._autoflush()
  File "/.venv/lib/python3.9/site-packages/sqlalchemy/orm/session.py", line 1633, in _autoflush
    util.raise_(e, with_traceback=sys.exc_info()[2])
  File "/.venv/lib/python3.9/site-packages/sqlalchemy/util/compat.py", line 182, in raise_
    raise exception
  File "/.venv/lib/python3.9/site-packages/sqlalchemy/orm/session.py", line 1622, in _autoflush
    self.flush()
  File "/.venv/lib/python3.9/site-packages/sqlalchemy/orm/session.py", line 2540, in flush
    self._flush(objects)
  File "/.venv/lib/python3.9/site-packages/sqlalchemy/orm/session.py", line 2682, in _flush
    transaction.rollback(_capture_exception=True)
  File "/.venv/lib/python3.9/site-packages/sqlalchemy/util/langhelpers.py", line 68, in __exit__
    compat.raise_(
  File "/.venv/lib/python3.9/site-packages/sqlalchemy/util/compat.py", line 182, in raise_
    raise exception
  File "/.venv/lib/python3.9/site-packages/sqlalchemy/orm/session.py", line 2642, in _flush
    flush_context.execute()
  File "/.venv/lib/python3.9/site-packages/sqlalchemy/orm/unitofwork.py", line 422, in execute
    rec.execute(self)
  File "/.venv/lib/python3.9/site-packages/sqlalchemy/orm/unitofwork.py", line 586, in execute
    persistence.save_obj(
  File "/.venv/lib/python3.9/site-packages/sqlalchemy/orm/persistence.py", line 239, in save_obj
    _emit_insert_statements(
  File "/.venv/lib/python3.9/site-packages/sqlalchemy/orm/persistence.py", line 1135, in _emit_insert_statements
    result = cached_connections[connection].execute(
  File "/.venv/lib/python3.9/site-packages/sqlalchemy/engine/base.py", line 1011, in execute
    return meth(self, multiparams, params)
  File "/.venv/lib/python3.9/site-packages/sqlalchemy/sql/elements.py", line 298, in _execute_on_connection
    return connection._execute_clauseelement(self, multiparams, params)
  File "/.venv/lib/python3.9/site-packages/sqlalchemy/engine/base.py", line 1124, in _execute_clauseelement
    ret = self._execute_context(
  File "/.venv/lib/python3.9/site-packages/sqlalchemy/engine/base.py", line 1316, in _execute_context
    self._handle_dbapi_exception(
  File "/.venv/lib/python3.9/site-packages/sqlalchemy/engine/base.py", line 1510, in _handle_dbapi_exception
    util.raise_(
  File "/.venv/lib/python3.9/site-packages/sqlalchemy/util/compat.py", line 182, in raise_
    raise exception
  File "/.venv/lib/python3.9/site-packages/sqlalchemy/engine/base.py", line 1276, in _execute_context
    self.dialect.do_execute(
  File "/.venv/lib/python3.9/site-packages/sqlalchemy/engine/default.py", line 609, in do_execute
    cursor.execute(statement, parameters)
sqlalchemy.exc.IntegrityError: (raised as a result of Query-invoked autoflush; consider using a session.no_autoflush block if this flush is occurring prematurely)
(sqlite3.IntegrityError) UNIQUE constraint failed: users.username
[SQL: INSERT INTO users (username, user_type_id) VALUES (?, ?)]
[parameters: ('1', 1)]
(Background on this error at: http://sqlalche.me/e/13/gkpj)
```

**Versions.**
 - OS: macOS Mojave 10.14.6 (18G95)
 - Python: 3.9.1
 - SQLAlchemy: 1.3.22
 - Database: SQLite 3.24.0 
 - DBAPI: sqlite3

**Additional context**

**Have a nice day!**

**Comment by zzzeek** [comment]
hey there -

the error message is trying to explain exactly the problem here:

"raised as a result of Query-invoked autoflush; consider using a session.no_autoflush block if this flush is occurring prematurely"

"no_autoflush" should be applied to eliminate the premature flush:

https://docs.sqlalchemy.org/en/13/orm/session_basics.html#flushing

additionally, the reason User is getting pushed into the Session before you'd want it to is due to this behavior:

https://docs.sqlalchemy.org/en/13/orm/cascades.html#controlling-cascade-on-backrefs

above, when you assign self._user_type = tmp_query, your User object gets placed into the Session when you didn't expect it to as UserType is persistent on the second run.  setting cascade_backrefs=False will prevent this.

the above issue is reversed in 1.4 and the default setting for "cascade_backrefs" is False.

**Comment by funk1d** [comment]
Maybe I missed something, but by setting `cascade_backrefs=False` in `relation` won't work:
```python
_user_type = relation('UserType',
                          order_by='UserType._name',
                          backref='_users', cascade_backrefs=False) # still get the sqlalchemy.exc.IntegrityError: (raised as a result of Query-invoked autoflush; consider using a session.no_autoflush block if this flush is occurring prematurely)
``` 
When trying the answer from [here](https://stackoverflow.com/a/35477608), it did not show the `autoflush` error when executing query, but actually it doesn't really fix the problem:
```python
from pathlib import Path

from sqlalchemy import Column, ForeignKey, Integer, String, create_engine
from sqlalchemy.engine.base import Engine
from sqlalchemy.ext.declarative import declarative_base
from sqlalchemy.orm import Session, relation

Base = declarative_base()
_engine: Engine = create_engine(
    'sqlite:///' + str(Path(__file__).parent.joinpath('db.sqlite3')))
db_session = Session(_engine, autoflush=False)  # set autoflush=False here


class User(Base):
    __tablename__ = 'users'
    _id = Column('id', Integer, primary_key=True)
    _username = Column('username', String, unique=True, nullable=False)
    _user_type_id = Column('user_type_id',
                           Integer,
                           ForeignKey('user_type.id'),
                           nullable=False)
    _user_type = relation('UserType',
                          order_by='UserType._name',
                          backref='_users',
                          cascade_backrefs=False) # set cascade_backrefs here

    def __init__(self, username: str, user_type: str):
        self._username = username
        tmp_query = db_session.query(UserType).filter_by(
            _name=user_type).first()
        if tmp_query:
            self._user_type = tmp_query
        else:
            self._user_type = UserType(user_type)


class UserType(Base):
    __tablename__ = 'user_type'
    _id = Column('id', Integer, primary_key=True)
    _name = Column('name', String, unique=True, nullable=False)

    def __init__(self, name: str):
        self._name = str(name)


Base.metadata.create_all(_engine)

if not db_session.query(User).filter_by(_username='1').first():
    db_session.add(User('1', '1'))
    db_session.commit()

x = User('1', '1')
x = User('1', '1') # all fine here

print(db_session.query(UserType).filter_by(_name='1').first()._users)
# [<__main__.User object at 0x10d74e460>, <__main__.User object at 0x10d0b3a00>, <__main__.User object at 0x10d5b7700>]
# having duplicate User objects here, which will cause the commit() failure below

db_session.add(User('2', '2'))
db_session.commit()
# sqlalchemy.exc.IntegrityError: (sqlite3.IntegrityError) UNIQUE constraint failed: users.username
# [SQL: INSERT INTO users (username, user_type_id) VALUES (?, ?)]
# [parameters: ('1', 1)]
# (Background on this error at: http://sqlalche.me/e/13/gkpj)
```

**Comment by zzzeek** [comment]
> Maybe I missed something, but by setting `cascade_backrefs=False` in `relation` won't work:

yes, the cascade_backrefs goes on the side that is receiving the event by which User becomes associated with UserType which is the "many" side, so it would be like this:

```
from sqlalchemy.orm import backref


class User(Base):
    __tablename__ = "users"
    _id = Column("id", Integer, primary_key=True)
    _username = Column("username", String, unique=True, nullable=False)
    _user_type_id = Column(
        "user_type_id", Integer, ForeignKey("user_type.id"), nullable=False
    )
    _user_type = relation(
        "UserType",
        order_by="UserType._name",
        backref=backref(
            "_users",
            cascade_backrefs=False,
        ),
    )  

```

> When trying the answer from [here](https://stackoverflow.com/a/35477608), it did not show the `autoflush` error when executing query, but actually it doesn't really fix the problem:

 db_session = Session(_engine, autoflush=False)  # set autoflush=False here

I wouldn't do it like that.  use the no_autoflush context manager as described in the error message and the document that I linked:

```
    def __init__(self, username: str, user_type: str):
        self._username = username
        with db_session.no_autoflush:
            tmp_query = (
                db_session.query(UserType).filter_by(_name=user_type).first()
            )
        if tmp_query:
            self._user_type = tmp_query
        else:
            self._user_type = UserType(user_type)

```

that will resolve the issue you describe, "unnecessary INSERT INTO when making a query".   

as far as why your program fails later, it's because you are making two User() objects with the same value for a column that has a UNIQUE constraint on it, adding them both to the Session due to the cascade behavior mentioned previously, and this rightfully causes a SQL error when you finally commit the transaction.

if you set up the cascade flags as above, you will *still* get a warning message:


```
[<__main__.User object at 0x7ffafc9361c0>, <__main__.User object at 0x7ffafcb9ffd0>, <__main__.User object at 0x7ffafc968e20>]
/home/classic/dev/sqlalchemy/lib/sqlalchemy/orm/unitofwork.py:301: SAWarning: Object of type <User> not in session, add operation along 'UserType._users' will not proceed
  util.warn(

```

this message refers to the fact that you created a UserType object that is persisted to the database and associated it with a User() object that was not in the Session, and this continues to not be clear as to your intent.

overall the issue is, why are you creating two User() objecs with conflicting data and associating them both with a UserType object from a single session/transaction.   This is really the source of all your issues.

**Comment by funk1d** [comment]
I actually fall into this issue by accidentally duplicating 2 create code of `User('1', '1')`, not on purpose.  
Maybe a dumb question, but I'm curious about: what is the best practice for this scenario?  
Since one *could* make mistake like "creating two User() objects with conflicting data and associating them both with a UserType object from a single session/transaction".  
And according to the document [When do I construct a Session, when do I commit it, and when do I close it?](https://docs.sqlalchemy.org/en/13/orm/session_basics.html#when-do-i-construct-a-session-when-do-i-commit-it-and-when-do-i-close-it), I shouldn't create new session inside the function to avoid the **single session/transaction**.
Should always check the model constraint in `__init__`?

**Comment by zzzeek** [comment, marked as answer]
> I actually fall into this issue by accidentally duplicating 2 create code of `User('1', '1')`, not on purpose.
> Maybe a dumb question, but I'm curious about: what is the best practice for this scenario?
> Since one _could_ make mistake like "creating two User() objects with conflicting data and associating them both with a UserType object from a single session/transaction".

sure, if you make that mistake, SQLAlchemy takes your intent to the database, which raises an integrity error.    Once people are fluent with the model that an ORM object represents a database row, this is a pretty straightforward error condition and it suggests the calling code needs to be adjusted to do a SELECT for that row first if they want to work with it, or otherwise adjust the approach being used depending on what it is you're trying to do.


> And according to the document [When do I construct a Session, when do I commit it, and when do I close it?](https://docs.sqlalchemy.org/en/13/orm/session_basics.html#when-do-i-construct-a-session-when-do-i-commit-it-and-when-do-i-close-it), I shouldn't create new session inside the function to avoid the **single session/transaction**.

you shouldn't use multiple sessions normally, that's true.  you would still have the two rows conflicting here.

> Should always check the model constraint in `__init__`?

i would not suggest that as it's wasteful.  organize the code so that it emits SELECT for rows which many exist already.   

this mostly depends on what you're trying to do.  if you have a program that's importing data from a set of keys where some may exist already, SELECT those keys up front and put them in a dictionary so you know those rows are there already.   Or you can use a method like [Session.merge()](https://docs.sqlalchemy.org/en/13/orm/session_state_management.html#unitofwork-merging), or if you want to be very SQL efficient you can skip the ORM for this particular operation and use an INSERT.. ON DUPLICATE kind of construct which we do support for SQLite in the SQLAlchemy 1.4 series.

what I usually would do for the "import a set of keys" use case is:

1. sort all the data that you want to insert in order of the "key" in question
2. work from the data being imported in batches of say 1000 objects each
3. for each batch, do a SELECT with an IN to get all the records in the DB that match for those records, put them in a dictionary keyed to the "key" in question
4. import your data by adding new objects for those not in the dict and updating those that are already in the dict
5. emit session.flush() after each batch.


the above assumes a lot about what you might be doing, but if you're having a case where you aren't sure if rows exist, that is usually what people are trying to do.
